### DISCUSSION TITLE
Feature request: 将 geo 数据文件的更新功能集成至 v2ctl / v2ray 工具集

### DISCUSSION THREAD

**Discussion by sixg0000d** [v2fly/v2ray-core#628] [Ideas]
\* `geoip.dat` 和 `geosite.dat `，在标题中称为 geo 数据文件，以下统称 `geofiles`

### 利
- 对于打包者：由于两个 geofiles 及 core 之间的更新周期不同，打包者要么需要将三者分开打包并对 geofiles 进行频繁升级打包，要么放弃对 geofiles 进行更新，直接将源码内的 geofiles 随 core 安装进指定目录。在集成该功能后，在不需要对 geofiles 进行额外打包的同时，还能让用户自行更新 geofiles 。
- 对于用户：将能更有选择性地、能自定义下载来源的更新 geofiles ，Linux 用户通过使用 `systemd.timer` ，将能自定义自动更新 geofiles 的周期。
- 对于 [v2fly/fhs-install-v2ray](https://github.com/v2fly/fhs-install-v2ray) ：可以不用再继续维护功能相同的 `install-dat-release.sh` ，成为一个单文件脚本项目。

### 害
- 可能会增加 v2ray 编译成的二进制文件体积

### 初步命令行设想
```
Usage:
v2ctl geoupdate [--geoip] [--geosite] [--path=<path>]
    --geoip  是否下载 geoip.dat
    --geosite  是否下载 geosite.dat
    --path <path>  将文件保存到<path>
```
> 按照 v5 版的修改方向，`v2ctl` 可能将集成至 `v2ray` ，固届时上面的命令行需改为 `v2ray geoupdate ...`

然后用户可以通过以下新的有效环境变量设定 geofiles 的下载地址：
```
V2RAY_GEOIP_URL // geoip.dat 下载地址
V2RAY_GEOIP_CHECKSUM_URL // geoip.dat 校验和下载地址，设为空字符串则跳过校验直接安装 / 升级
V2RAY_GEOSITE_URL // geosite.dat 下载地址
V2RAY_GEOSITE_CHECKSUM_URL // geosite.dat 校验和下载地址，设为空字符串则跳过校验直接安装 / 升级
```
用户可以通过已有的有效环境变量 `V2RAY_LOCATION_ASSET` 设置 geofiles 的保存位置，但若存在 `--path` 参数将会被覆盖

### 补充
1. 为什么我想要把下载地址各 URL 的设置放在环境变量里，因为我觉得一般用户运行命令行工具，手动输入不会想要输入这么长的参数，但是加入配置文件感觉也没必要，而且 `systemd.exec` 能完美的调整环境变量，于是就想用这种方法自定义下载地址。但是说实话我并不了解 Windows 上能否使用环境变量和适用程度，所以还请各位指正。
2. 所以如果按照我的想法设计，除代码外，则还需要添加以上 4 个新的有效环境变量和相关文档

### 用例
```bash
# case 1
v2ctl geoupdate
```
什么都不会做

```bash
# case 2
v2ctl geoupdate --geoip
```
将会下载 [geoip.dat](https://github.com/v2fly/geoip/releases/latest/download/geoip.dat) 并下载 [geoip.dat.sha256sum](https://github.com/v2fly/geoip/releases/latest/download/geoip.dat.sha256sum) 获取校验和以验证文件完整型，成功后保存到当前目录 (workdir) 。

```bash
# case 3
V2RAY_LOCATION_ASSET="/usr/local/share/v2ray" \
V2RAY_GEOIP_URL="https://github.com/Loyalsoldier/geoip/releases/latest/download/geoip.dat" \
V2RAY_GEOIP_CHECKSUM_URL="" \
v2ctl geoupdate --geoip
```
将会下载 [geoip.dat](https://github.com/Loyalsoldier/geoip/releases/latest/download/geoip.dat) ，不做验证，直接保存到 `/usr/local/share/v2ray` 。

```bash
# case 4
V2RAY_LOCATION_ASSET="/usr/local/share/v2ray" \
v2ctl geoupdate --geoip --geosite --path ~/v2ray
```
将会下载 [geoip.dat](https://github.com/v2fly/geoip/releases/latest/download/geoip.dat) 和 [dlc.dat](https://github.com/v2fly/domain-list-community/releases/latest/download/dlc.dat) 并下载 [geoip.dat.sha256sum](https://github.com/v2fly/geoip/releases/latest/download/geoip.dat.sha256sum) 和 [dlc.dat.sha256sum](https://github.com/v2fly/domain-list-community/releases/latest/download/dlc.dat.sha256sum) 获取校验和以验证文件完整型，成功后保存到家目录下的 v2ray 文件夹中 。

由于 geofiles 不存在版本号所以我觉得下载时保持时间戳有一定的必要性。并且我发现 github raw 中下载文件的响应由于没有 `Last-Modified` 头部所以无法做到保持时间戳，这也是我选择用 github release download 作为默认下载地址的原因。

**Comment by CalmLong** [comment]
~~都把 `v2ctl` 合进去了，还会在乎这点体积吗~~

**Comment by CalmLong** [comment]
不过有个比较现实的情况就是有不少人用的第三方的 geo 资源，应该可以支持自定义 `V2RAY_GEOIP_URL`

**Comment by sixg0000d** [comment]
刚才想到了一个问题，那就是下载加速，Github上的资源在中国国内访问可能会很慢，除非让用户自行使用镜像，否则`geoupdate`需要做对下载文件进行代理的处理？

**Comment by kslr** [comment]
除了下载速度这点，发布的包里是包含数据文件的，因此这可能只是一个简单更新功能。

ip 的更新速度是每周一次，site 取决于更改，这个频率比较高点，但这些基本和发布频率是保持一致的。

不过如果你想做这个功能，也完全没有问题
